### PR TITLE
aws_cur_report_definition - Parquet, Athena and more

### PR DESCRIPTION
**Pull request by robbruce**
This change updates aws_cur_report_definition resource and data source to match options available on AWS.

Added:

- Parquet Format
- Athena Additional Artifact
- Refresh Closed Reports
- Report Versioning

When using the AWS cur API, it very unhelpfully returns an empty ValidationException when certain combinations of options conflict with each other. The allowed combination of options has been added as part of the create step. For example, if Athena is being used the output format and compression *must* be Parquet and the s3 prefix *cannot* be empty!

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #8765

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Add Parquet Format, Athena Additional Artifact, Refresh Closed Reports and Report Versioning to aws_cur_report_definition resource and data source
```

Some slight changes were made on the acceptance tests because some race conditions can occur when the bucket policy had not finished applying (even with a depends_on) and so a ValidationException occurs.

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAwsCurReportDefinition'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAwsCurReportDefinition -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsCurReportDefinition_basic
=== PAUSE TestAccAwsCurReportDefinition_basic
=== RUN   TestAccAwsCurReportDefinition_parquet
=== PAUSE TestAccAwsCurReportDefinition_parquet
=== RUN   TestAccAwsCurReportDefinition_athena
=== PAUSE TestAccAwsCurReportDefinition_athena
=== RUN   TestAccAwsCurReportDefinition_refresh
=== PAUSE TestAccAwsCurReportDefinition_refresh
=== RUN   TestAccAwsCurReportDefinition_overwrite
=== PAUSE TestAccAwsCurReportDefinition_overwrite
=== CONT  TestAccAwsCurReportDefinition_basic
=== CONT  TestAccAwsCurReportDefinition_refresh
=== CONT  TestAccAwsCurReportDefinition_parquet
=== CONT  TestAccAwsCurReportDefinition_overwrite
=== CONT  TestAccAwsCurReportDefinition_athena
--- PASS: TestAccAwsCurReportDefinition_overwrite (80.17s)
--- PASS: TestAccAwsCurReportDefinition_basic (80.65s)
--- PASS: TestAccAwsCurReportDefinition_refresh (80.78s)
--- PASS: TestAccAwsCurReportDefinition_parquet (80.90s)
--- PASS: TestAccAwsCurReportDefinition_athena (83.89s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	85.611s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap	0.560s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags	0.268s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/naming	0.532s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/batch/equivalency	0.179s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/eks/token	0.250s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes	0.495s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSAT001	0.350s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR001	0.150s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/fmtsprintfcallexpr	0.106s [no tests to run]
```

```
$ make testacc TESTARGS='-run=TestAccDataSourceAwsCurReportDefinition'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccDataSourceAwsCurReportDefinition -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccDataSourceAwsCurReportDefinition_basic
=== PAUSE TestAccDataSourceAwsCurReportDefinition_basic
=== CONT  TestAccDataSourceAwsCurReportDefinition_basic
--- PASS: TestAccDataSourceAwsCurReportDefinition_basic (89.42s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	91.195s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap	0.657s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags	0.342s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/naming	0.423s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/batch/equivalency	0.164s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/eks/token	0.259s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes	0.370s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSAT001	0.457s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR001	0.126s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/fmtsprintfcallexpr	0.229s [no tests to run]
```

### CODE REVIEW

**Comment by robbruce** [comment]
Very similar to https://github.com/terraform-providers/terraform-provider-aws/pull/11567, but includes refresh and version options as well as data resource.

Didn't know 11567 existed until after creating the pull request!

**Comment by andymoore-synamedia** [comment]
currently using this in a stage env to test (in the hope this gets merged some time in the future and we can go back to proper master...)

hit a couple of things when trying to use, might be worth adding them to the docs:

```
When ATHENA exists within additional_artifacts, no other artifact type can be declared
When ATHENA exists within additional_artifacts, report_versioning must be OVERWRITE_REPORT```

**Comment by andymoore-synamedia** [comment]
👍

**Comment by sckevmit** [comment]
Can we please get this merged?

**Comment by robbruce** [comment]
Updated and re-ran acceptance tests:

```
make testacc TESTARGS='-run=TestAccAwsCurReportDefinition'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsCurReportDefinition -timeout 120m
=== RUN   TestAccAwsCurReportDefinition_basic
=== PAUSE TestAccAwsCurReportDefinition_basic
=== RUN   TestAccAwsCurReportDefinition_parquet
=== PAUSE TestAccAwsCurReportDefinition_parquet
=== RUN   TestAccAwsCurReportDefinition_athena
=== PAUSE TestAccAwsCurReportDefinition_athena
=== RUN   TestAccAwsCurReportDefinition_refresh
=== PAUSE TestAccAwsCurReportDefinition_refresh
=== RUN   TestAccAwsCurReportDefinition_overwrite
=== PAUSE TestAccAwsCurReportDefinition_overwrite
=== CONT  TestAccAwsCurReportDefinition_basic
=== CONT  TestAccAwsCurReportDefinition_athena
=== CONT  TestAccAwsCurReportDefinition_refresh
=== CONT  TestAccAwsCurReportDefinition_overwrite
=== CONT  TestAccAwsCurReportDefinition_parquet
--- PASS: TestAccAwsCurReportDefinition_refresh (38.38s)
--- PASS: TestAccAwsCurReportDefinition_overwrite (38.89s)
--- PASS: TestAccAwsCurReportDefinition_athena (38.90s)
--- PASS: TestAccAwsCurReportDefinition_parquet (38.92s)
--- PASS: TestAccAwsCurReportDefinition_basic (39.53s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	41.293s
```

```
make testacc TESTARGS='-run=TestAccDataSourceAwsCurReportDefinition'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsCurReportDefinition -timeout 120m
=== RUN   TestAccDataSourceAwsCurReportDefinition_basic
=== PAUSE TestAccDataSourceAwsCurReportDefinition_basic
=== CONT  TestAccDataSourceAwsCurReportDefinition_basic
--- PASS: TestAccDataSourceAwsCurReportDefinition_basic (43.24s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	44.900s
```

**Comment by ewbankkit** [comment]
@robbruce Thanks for the contribution.
A couple of (relatively minor) suggestions:

* Replace explicit list of values in the `ValidateFunc`s with AWS SDK `_Values()`: #14601
* Could you split the acceptance test configuration generating function to keep `_basic` as it was and add a second one that takes all the new options you added. This will help ensure that there are no regressions

**Comment by robbruce** [comment]
@ewbankkit changes requested have been implemented.

Note, this also fixes #14827 (a bug introduced by #14432)

Unit tests re-ran (had to reduce parallel tests as an AWS limit was reached on our account as other reports already existed!)

```
make testacc TESTARGS='-run=TestAccAwsCurReportDefinition' ACCTEST_PARALLELISM=1                                                                                                                                         
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAwsCurReportDefinition -timeout 120m
=== RUN   TestAccAwsCurReportDefinition_basic
=== PAUSE TestAccAwsCurReportDefinition_basic
=== RUN   TestAccAwsCurReportDefinition_textOrCsv
=== PAUSE TestAccAwsCurReportDefinition_textOrCsv
=== RUN   TestAccAwsCurReportDefinition_parquet
=== PAUSE TestAccAwsCurReportDefinition_parquet
=== RUN   TestAccAwsCurReportDefinition_athena
=== PAUSE TestAccAwsCurReportDefinition_athena
=== RUN   TestAccAwsCurReportDefinition_refresh
=== PAUSE TestAccAwsCurReportDefinition_refresh
=== RUN   TestAccAwsCurReportDefinition_overwrite
=== PAUSE TestAccAwsCurReportDefinition_overwrite
=== CONT  TestAccAwsCurReportDefinition_basic
--- PASS: TestAccAwsCurReportDefinition_basic (47.10s)
=== CONT  TestAccAwsCurReportDefinition_refresh
--- PASS: TestAccAwsCurReportDefinition_refresh (45.45s)
=== CONT  TestAccAwsCurReportDefinition_overwrite
--- PASS: TestAccAwsCurReportDefinition_overwrite (45.52s)
=== CONT  TestAccAwsCurReportDefinition_parquet
--- PASS: TestAccAwsCurReportDefinition_parquet (47.17s)
=== CONT  TestAccAwsCurReportDefinition_athena
--- PASS: TestAccAwsCurReportDefinition_athena (45.53s)
=== CONT  TestAccAwsCurReportDefinition_textOrCsv
--- PASS: TestAccAwsCurReportDefinition_textOrCsv (47.58s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	280.045s
```

```
make testacc TESTARGS='-run=TestAccDataSourceAwsCurReportDefinition' ACCTEST_PARALLELISM=1                                                                                                                               
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccDataSourceAwsCurReportDefinition -timeout 120m
=== RUN   TestAccDataSourceAwsCurReportDefinition_basic
=== PAUSE TestAccDataSourceAwsCurReportDefinition_basic
=== RUN   TestAccDataSourceAwsCurReportDefinition_additional
=== PAUSE TestAccDataSourceAwsCurReportDefinition_additional
=== CONT  TestAccDataSourceAwsCurReportDefinition_basic
--- PASS: TestAccDataSourceAwsCurReportDefinition_basic (88.86s)
=== CONT  TestAccDataSourceAwsCurReportDefinition_additional
--- PASS: TestAccDataSourceAwsCurReportDefinition_additional (137.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	228.578s
```

File: aws/resource_aws_cur_report_definition.go [hunk 1]
```diff
@@ -81,34 +84,129 @@ func resourceAwsCurReportDefinition() *schema.Resource {
...
+
+	if !hasAthena && len(additionalArtifacts) > 1 && (*format == costandusagereportservice.ReportFormatParquet) {
+		return fmt.Errorf(
+			"When additional_artifacts includes %s and/or %s, format must not be %s",
+			costandusagereportservice.AdditionalArtifactQuicksight,
+			costandusagereportservice.AdditionalArtifactRedshift,
+			costandusagereportservice.ReportFormatParquet,
+		)
+	}
+	// end checks
```
Comment (ewbankkit) [inline aws/resource_aws_cur_report_definition.go:194]: Minor suggestion to move this logic into a separate function that can the be unit tested.

Comment (robbruce) [reply aws/resource_aws_cur_report_definition.go:194]: @ewbankkit makes sense

**Comment by robbruce** [comment]
@ewbankkit just as well you mentioned that as the unit test did highlight a bug

```go
if !hasAthena && len(additionalArtifacts) > 1 && (*format == costandusagereportservice.ReportFormatParquet) {
```

should have been!
```go
if !hasAthena && len(additionalArtifacts) > 0 && (*format == costandusagereportservice.ReportFormatParquet) {
```

**Review by bflad — APPROVED** [review]
Thank you for these updates, @robbruce 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAwsCurReportDefinition_overwrite (37.09s)
--- PASS: TestAccAwsCurReportDefinition_athena (37.35s)
--- PASS: TestAccAwsCurReportDefinition_refresh (38.27s)
--- PASS: TestAccAwsCurReportDefinition_textOrCsv (38.99s)
--- PASS: TestAccAwsCurReportDefinition_basic (39.75s)
--- PASS: TestAccAwsCurReportDefinition_parquet (40.34s)

--- PASS: TestAccDataSourceAwsCurReportDefinition_basic (40.55s)
--- PASS: TestAccDataSourceAwsCurReportDefinition_additional (40.62s)
```

**Pull request merged**